File: lib/models/supplier.dart
class Supplier {
  final int supplierId;
  final String supplierName;
  final String? address;
  final String? contactInfo;
  final int? leadTimeDays;
  final int? createdBy;

  Supplier({
    required this.supplierId,
    required this.supplierName,
    this.address,
    this.contactInfo,
    this.leadTimeDays,
    this.createdBy,
  });

  factory Supplier.fromJson(Map<String, dynamic> json) {
    int? intValue(dynamic value) {
      if (value is num) {
        return value.toInt();
      }
      return int.tryParse(value?.toString() ?? '');
    }

    final contactInfo = (json['contact_info']?.toString().isNotEmpty ?? false)
        ? json['contact_info']?.toString()
        : (json['phone']?.toString().isNotEmpty ?? false)
            ? json['phone']?.toString()
            : json['email']?.toString();

    return Supplier(
      supplierId: intValue(json['supplier_id'] ?? json['id']) ?? 0,
      supplierName: (json['supplier_name'] ?? json['name'] ?? '').toString(),
      address: json['address']?.toString(),
      contactInfo: contactInfo,
      leadTimeDays: intValue(json['lead_time_days']),
      createdBy: intValue(json['created_by']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'supplier_id': supplierId,
      'supplier_name': supplierName,
      'address': address,
      'contact_info': contactInfo,
      'lead_time_days': leadTimeDays,
      'created_by': createdBy,
    };
  }
}

File: app/models/supplier.py
from sqlalchemy import ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class Supplier(Base):
    __tablename__ = "supplier"

    id: Mapped[int] = mapped_column("supplier_id", Integer, primary_key=True, index=True)
    name: Mapped[str] = mapped_column("supplier_name", String(100), unique=True, nullable=False, index=True)
    contact_info: Mapped[str | None] = mapped_column(String(150), nullable=True)
    lead_time_days: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    address: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_by: Mapped[int] = mapped_column(Integer, ForeignKey("user.user_id"), nullable=False, index=True)

    products: Mapped[list["Product"]] = relationship("Product", back_populates="supplier")
    stock_receipts: Mapped[list["StockReceipt"]] = relationship("StockReceipt", back_populates="supplier")

    @property
    def email(self) -> str | None:
        value = (self.contact_info or "").strip()
        return value if "@" in value else None

    @property
    def phone(self) -> str | None:
        value = (self.contact_info or "").strip()
        return value if value and "@" not in value else None

    @property
    def is_active(self) -> bool:
        return True
